src/collapse.ts

```diff
@@ -0,0 +1,95 @@
+import assert from 'assert';
+
+export default class Collapse {
+    private _horizontal: boolean;
+    private _property: 'width' | 'height';
+    private _visible: boolean = false;
+
+    constructor(private _element: HTMLElement) {
+        this._horizontal = this._element.classList.contains('horizontal');
+        this._property = this._horizontal ? 'width' : 'height';
+
+        this._element.addEventListener('transitionend', (event) => {
+            if (event.target === event.currentTarget && event.propertyName === this._property) {
+                this._element.classList.replace('collapsing', 'collapse');
+
+                if (this._visible) {
+                    this._element.classList.add('show');
+                    this._element.style.removeProperty(this._property);
+                }
+            }
+        });
+    }
+
+    hide() {
+        const size = this._element.getBoundingClientRect()[this._property];
+        this._element.style.setProperty(this._property, `${size}px`);
+
+        this._element.classList.replace('collapse', 'collapsing');
+        this._element.classList.remove('show');
+
+        // eslint-disable-next-line no-unused-expressions
+        this._element.offsetHeight;
+        this._element.style.removeProperty(this._property);
+
+        this._visible = false;
+    }
+
+    show() {
+        const scrollProperty = this._horizontal ? 'scrollWidth' : 'scrollHeight';
+
+        this._element.classList.replace('collapse', 'collapsing');
+        this._element.style.setProperty(this._property, '0');
+
+        const size = this._element[scrollProperty];
+        this._element.style.setProperty(this._property, `${size}px`);
+
+        this._visible = true;
+    }
+
+    toggle(value: boolean = !this._visible) {
+        if (value) {
+            this.show();
+        } else {
+            this.hide();
+        }
+    }
+
+    addTrigger(el: HTMLElement) {
+        el.addEventListener('click', (event) => {
+            event.preventDefault();
+            this.toggle();
+        });
+    }
+
+    static initialize(root: HTMLElement = document.body) {
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+        for (const el of root.querySelectorAll('[data-bs-toggle]')) {
+            if (el instanceof HTMLElement && 'bsTarget' in el.dataset) {
+                const target: HTMLElement | null = root.querySelector(
+                    (el.dataset as { bsTarget: string }).bsTarget
+                );
+
+                if (target) {
+                    getCollapse(target).addTrigger(el);
+                }
+            }
+        }
+    }
+}
+
+const map = new WeakMap<HTMLElement, Collapse>();
```

Comment:
could you explain why a weak map is required here? and what the function below does/why it is required?

Reply:
The point of `getCollapse()` is to access existing `Collapse` instances without the reference to that instance but instead one to the corresponding element.

```js
// Initially
let collapse = getCollapse(element);

// Later (same collapse)
let collapse = getCollapse(element);

// Removing the element
element.remove(); // -> `collapse` will eventually be garbage-collected as the WeakMap does not count as a reference
```

Setting `element.collapse = new Collapse()` works fine, but pollutes the element object (using symbols would be an alternative though).

Reply:
Also `getFromMap()` underneath is just a way of generalizing this solution to any key and value given a `WeakMap` and a function to generate new values when missing from the map.

Reply:
ok, that sounds good!